src/backend/storage/data_table.cpp

```diff
@@ -1009,5 +1018,468 @@ column_map_type DataTable::GetStaticColumnMap(const std::string &table_name,
+    column_list.push_back(schema->GetColumn(table_column_id));
+    sample_schema_list.push_back(catalog::Schema(column_list));
+  }
+
+  return;
+}
+
+/*
+ * BuildSampleTileGroup() - Build a tile group instance with sampling
+ *                          arguments
+ *
+ * It uses the sampling table's column map (pure columnar), schema list
+ * (one column per schema), and tuple count (must take sample first)
+ *
+ * NOTE: This function allocates new tile group ID and assign it to the
+ * created tile group, and also assign it to class member variable
+ */
+TileGroup *DataTable::BuildSampleTileGroup() {
+  if(GetOptimizerSampleSize() == 0LU) {
+    LOG_INFO("Sample size is zero. Please take samples first");
+
+    return nullptr;
+  }
+
+  // Allocate new tile group ID, and assign to class member
+  sampled_tile_group_id = catalog::Manager::GetInstance().GetNextOid();
+
+  // Create a tile group with column map, schema list and tuple count
+  // of the sample table rather than data table
+  TileGroup *tile_group = TileGroupFactory::GetTileGroup(
+      database_oid,            // Stored in data table
+      table_oid,               // Stored in data table
+      sampled_tile_group_id,   // Allocated just now
+      this,                    // Table reference
+      sample_schema_list,      // Used by TileGroup to build its internal tile list
+      sample_column_map,       // Used by TileGroup to locate table column
+      GetOptimizerSampleSize() // Used by TileGroup to allocate internal storage
+      );
+
+  return tile_group;
+}
+
+/*
+ * FillSampleTileGroup() - Fill sample tile group created elsewhere
+ *                         with actual tuples
+ *
+ * We always store samples in columnar format
+ *
+ * NOTE 1: Currently we do not write data through the TileGroup interface
+ * which means we directly go to Tile interface and GetValue()/SetValue()
+ * This is really bad because it circumvents all controls provided by TileGroup
+ * including MVCC
+ *
+ * NOTE 2: Despite the statements above, we do make use of GetNextEmptyTupleSlot()
+ * to keep TileGroupHeader consistent with the sampling table
+ */
+void DataTable::FillSampleTileGroup() {
+  if(GetOptimizerSampleSize() == 0LU) {
+    LOG_INFO("Sample has not been taken");
+
+    return;
+  }
+
+  // Make sure we have already materialized it
+  assert(sampled_tile_group_id != INVALID_OID);
+
+  // Get tile group for sampling table outside of
+  // the loop
+  auto sample_tile_group_p = GetSampleTileGroup();
+
+  // Iterate through our sampled pointers
+  for(ItemPointer &item : samples_for_optimizer) {
+    // Offset of tile group (block offset)
+    oid_t tile_group_offset = item.block;
+    // Offset inside tile group (row offset)
+    oid_t row_offset = item.offset;
+
+    auto tile_group_p = GetTileGroup(tile_group_offset);
+
+    // For each sampled row, we assign a row ID in the sample table
+    oid_t next_sample_row_id = sample_tile_group_p->GetHeader()->GetNextEmptyTupleSlot();
```

Comment:
next_sample_row_id could be invalid, when we run out of the free tuples in sample_tile_group